scripts: ci: introduce soc name check in check_compliance

tejlmand opened the pull request:
soc.yml files define SoC names which are used in board.yml. All SoC names and directories are exported to the build system and can be referenced using the SoC name as identifier.

Kconfig defines a CONFIG_SOC setting with the same name which can be used in build system and is selected by the board.

Thus the CONFIG_SOC value can be used to lookup the details of the SoC.

This commit introduces a new compliance check which ensures the SoC name and the CONFIG_SOC name value are in sync.

--------

Failures in this PR is being addressed here:
- #69308
- #69322

tejlmand committed:
scripts: ci: introduce soc name check in check_compliance

soc.yml files define SoC names which are used in board.yml.
All SoC names and directories are exported to the build system and
can be referenced using the SoC name as identifier.

Kconfig defines a CONFIG_SOC setting with the same name which can be
used in build system and is selected by the board.

Thus the CONFIG_SOC value can be used to lookup the details of the SoC.

This commit introduces a new compliance check which ensures the SoC name
and the CONFIG_SOC name value are in sync.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
This PR is expected to fail as it exposes mismatches which currently exists in collab-hwm.
Ref: #69308

stephanosio commented on scripts/ci/check_compliance.py:
```diff
@@ -658,6 +659,32 @@ def check_no_undef_within_kconfig(self, kconf):
+
+        soc_kconfig_names = set()
+        for node in kconf.node_iter():
+            if isinstance(node.item, kconfiglib.Symbol) and node.item.name == "SOC":
```
might want to add `# pylint: disable=undefined-variable` to suppress the complaint by pylint.

tejlmand replied:
Done.

57300 commented:
Should `CONFIG_SOC_SERIES` and `CONFIG_SOC_FAMILY` be checked as well?

By the way, did you ever consider having `CONFIG_SOC` inherited from CMake, just like `CONFIG_BOARD`? (based on board identifier and/or `list_hardware.py` output)

tejlmand commented:
> Should `CONFIG_SOC_SERIES` and `CONFIG_SOC_FAMILY` be checked as well?

Good points, could be nice to check those as well, although the immediate need is to ensure the SoC names are in sync because we use those vaules to sync the soc dir between list_hardware, CMake and Kconfig settings.
 
> By the way, did you ever consider having `CONFIG_SOC` inherited from CMake, just like `CONFIG_BOARD`? (based on board identifier and/or `list_hardware.py` output)

I did, but because during implementation of hwmv2 various adjustments has been made along the way.
Initially the first value after `/<identifier>` could be either a SoC or a variant (has later been fixed to soc).
Also SiP support has been mentioned during arch wg, currently a SiP must be modeled as a SoC, but the model is extensible so later we might want to support SiP, which can then support multiple SoCs.

In fact, the nRF9161is actually a SiP containing a nRF9120 SoC.
https://github.com/zephyrproject-rtos/zephyr/blob/d6cb9c967227b05e59a3f8e5a619fa695c174594/soc/nordic_nrf/nrf91/Kconfig.soc#L31-L32

Thus, for time being I've decided to not pass in the SoC, but instead use the existing Kconfig scheme of the board selecting the SoC, which we can later retrieve.

57300 commented on scripts/ci/check_compliance.py:
```diff
@@ -658,6 +659,32 @@ def check_no_undef_within_kconfig(self, kconf):
+                    soc_kconfig_names.add(d[0].name)
+
+        soc_name_warnings = []
+        for name in soc_names:
+            if name not in soc_kconfig_names:
+                soc_name_warnings.append(f"soc name: {name} not found in CONFIG_SOC defaults.")
```
This loops through `soc_names - soc_kconfig_names`, but I think `soc_kconfig_names - soc_names` should be checked as well, to ensure that the two sets are equal.

57300 replied:
Gotcha. I just thought the main purpose was to catch things like #69308, for which you'd have to test the opposite.

tejlmand replied:
I actually caught those with current implementation 😉

57300 replied:
Fair enough :) In this case, it was possible to figure out the problem from either direction.

tejlmand committed:
soc: emsdsp: align SoC name defined in soc.yml to Kconfig SOC setting

This commit aligns the SoC name for emsdp_em4 SoC defined in Kconfig.soc
to to name in soc.yml

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

abrodkin reviewed (approved):
Good catch and thanks for the fix!
